master/internal/db/postgres_experiments.go

```diff
@@ -252,3 +260,126 @@ ORDER BY v.end_time;`, &rows, metricName, experimentID, batchesProcessed, startT
 
 	return trials, endTime, nil
 }
+
+// TopTrialsByMetric chooses the subset of trials from an experiment that recorded the best values
+// for the specified metric at any point during the trial.
+func (db *PgDB) TopTrialsByMetric(experimentID int, maxTrials int, metric string,
+	smallerIsBetter bool) (trials []int32, err error) {
+	order := desc
+	aggregate := max
+	if smallerIsBetter {
+		order = asc
+		aggregate = min
+	}
+	err = db.sql.Select(&trials, fmt.Sprintf(`
+SELECT t.id FROM (
+  SELECT t.id,
+    %s((v.metrics->'validation_metrics'->$1)::text::numeric) as best_metric
+  FROM trials t
+    INNER JOIN steps s ON t.id=s.trial_id
+    RIGHT JOIN validations v ON s.id=v.step_id AND s.trial_id=v.trial_id
+  WHERE t.experiment_id=$2
+    AND v.state = 'COMPLETED'
+  GROUP BY t.id
+  ORDER BY best_metric %s
+  LIMIT $3
+) t;`, aggregate, order), metric, experimentID, maxTrials)
+	return trials, err
+}
+
+// TopTrialsByTrainingLength chooses the subset of trials that has been training for the highest
+// number of batches, using the specified metric as a tie breaker.
+func (db *PgDB) TopTrialsByTrainingLength(experimentID int, maxTrials int, metric string,
+	smallerIsBetter bool) (trials []int32, err error) {
+	order := desc
+	aggregate := max
+	if smallerIsBetter {
+		order = asc
+		aggregate = min
+	}
+
+	err = db.sql.Select(&trials, fmt.Sprintf(`
+SELECT t.id FROM (
+  SELECT t.id,
+    max(s.prior_batches_processed) as progress,
+    %s((v.metrics->'validation_metrics'->$1)::text::numeric) as best_metric
+  FROM trials t
+    INNER JOIN steps s ON t.id=s.trial_id
+    RIGHT JOIN validations v ON s.id=v.step_id AND s.trial_id=v.trial_id
+  WHERE t.experiment_id=$2
+    AND v.state = 'COMPLETED'
+  GROUP BY t.id
+  ORDER BY progress DESC, best_metric %s
+  LIMIT $3
+) t;`, aggregate, order), metric, experimentID, maxTrials)
+	return trials, err
+}
+
+type metricsSeriesWrapper struct {
+	Batches int       `db:"batches" json:"batches"`
```

Comment:
non-blocking: I dont think we need the json tags?

Reply:
Ah yes - I think that might be leftover from when I thought I could use one object all the way through without the need to do any intermediate conversion.